lib/ipam/ipam_test.go

```diff
@@ -12,88 +12,62 @@
-// Test cases (ReleaseIPs):
-// Test 1: release an IP that's not configured in any pools - expect a slice with the same IP as unallocatedIPs and no error.
-// Test 2: release an IP that's not allocated in the pool - expect a slice with one (unallocatedIPs) and no error.
-// Test 3: Assign 1 IPv4 with AssignIP from a configured pool and then release it.
-// - Assign should not return an error.
-// - ReleaseIP should return empty slice of IPs (unallocatedIPs) and no error.
-// Test 4: Assign 66 IPs (across 2 blocks) with AutoAssign from a configured pool then release them.
-// - Assign should not return an error.
-// - ReleaseIPs should return an empty slice of IPs (unallocatedIPs) and no error.
-// Test 5: Assign 1 IPv4 address with AssignIP then try to release 2 IPs.
-// - Assign should not return no error.
-// - ReleaseIPs should return a slice with one (unallocatedIPs) and no error.
-
-// Test cases (ClaimAffinity):
-// Test 1: claim affinity for an unclaimed IPNet of size 64 - expect 1 claimed blocks, 0 failed and expect no error.
-// Test 2: claim affinity for an unclaimed IPNet of size smaller than 64 - expect 0 claimed blocks, 0 failed and expect an error error.
-// Test 3: claim affinity for a IPNet that has an IP already assigned to another host.
-// - Assign an IP with AssignIP to "host-A" from a configured pool - expect 0 claimed blocks, 0 failed and expect no error.
-// - Claim affinity for "Host-B" to the block that IP belongs to - expect 3 claimed blocks and 1 failed.
-// Test 4: claim affinity to a block twice from different hosts.
-// - Claim affinity to an unclaimed block for "Host-A" - expect 4 claimed blocks, 0 failed and expect no error.
-// - Claim affinity to the same block again but for "host-B" this time - expect 0 claimed blocks, 4 failed and expect no error.
-
-package client_test
+package ipam
 
 import (
+	"context"
 	"errors"
 	"fmt"
-	"log"
 	"net"
+	"sort"
 
 	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
 	. "github.com/onsi/gomega"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/libcalico-go/lib/apiconfig"
 	"github.com/projectcalico/libcalico-go/lib/backend"
+	bapi "github.com/projectcalico/libcalico-go/lib/backend/api"
 	"github.com/projectcalico/libcalico-go/lib/backend/model"
-
-	. "github.com/onsi/ginkgo/extensions/table"
-
-	"github.com/projectcalico/libcalico-go/lib/api"
-	"github.com/projectcalico/libcalico-go/lib/client"
 	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
 	cnet "github.com/projectcalico/libcalico-go/lib/net"
 	"github.com/projectcalico/libcalico-go/lib/testutils"
 )
 
+// Implement an IP pools accessor for the IPAM client.
+type ipPoolAccessor struct {
+	pools map[string]bool
+}
+
+func (i *ipPoolAccessor) GetEnabledPools(ipVersion int) ([]cnet.IPNet, error) {
+	sorted := []string{}
+	// Get a sorted list of enabled pool CIDR strings.
+	for p, e := range i.pools {
+		if e {
+			sorted = append(sorted, p)
+		}
+	}
+	sort.Strings(sorted)
```

Comment:
not sure I understand the need for sorting here. Can you explain?

Reply:
Just to mimic what etcd will do and to minimize the changes required for the UTs (since they made the assumption that the pools were ordered).

Reply:
Oh yes - and to allow the tests to be deterministic.

Reply:
I'll add a comment.